// sanity/lib/cjs/form/inputs/BooleanInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BooleanInput = BooleanInput;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ui = require("@sanity/ui");

var _FormFieldHeaderText = require("../components/formField/FormFieldHeaderText");

var _FormFieldStatus = require("../components/formField/FormFieldStatus");

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

const CenterAlignedBox = (0, _styledComponents.default)(_ui.Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  align-self: center;\n"])));
const ZeroLineHeightBox = (0, _styledComponents.default)(_ui.Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  line-height: 0;\n"])));

function BooleanInput(props) {
  var _schemaType$options;

  const {
    id,
    value,
    schemaType,
    readOnly,
    elementProps
  } = props;
  const layout = ((_schemaType$options = schemaType.options) === null || _schemaType$options === void 0 ? void 0 : _schemaType$options.layout) || 'switch';
  const indeterminate = typeof value !== 'boolean';
  const checked = value || false;
  const LayoutSpecificInput = layout === 'checkbox' ? _ui.Checkbox : _ui.Switch;
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    border: true,
    radius: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(ZeroLineHeightBox, {
    padding: 3
  }, /*#__PURE__*/_react.default.createElement(LayoutSpecificInput, _extends({
    label: schemaType.title
  }, elementProps, {
    indeterminate: indeterminate,
    checked: checked,
    style: {
      margin: -4
    },
    disabled: readOnly
  }))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    paddingY: 3
  }, /*#__PURE__*/_react.default.createElement(_FormFieldHeaderText.FormFieldHeaderText, {
    description: schemaType.description,
    inputId: id // validation={validation}
    ,
    title: schemaType.title
  })), /*#__PURE__*/_react.default.createElement(CenterAlignedBox, {
    paddingX: 3,
    paddingY: 1
  }, /*#__PURE__*/_react.default.createElement(_FormFieldStatus.FormFieldStatus, {
    maxAvatars: 1,
    position: "top"
  }))));
}
