// sanity/lib/cjs/form/members/items/ArrayOfPrimitivesItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayOfPrimitivesItem = ArrayOfPrimitivesItem;

var _react = _interopRequireWildcard(require("react"));

var _useDidUpdate = require("../../hooks/useDidUpdate");

var _getEmptyValue = require("../../inputs/arrays/ArrayOfPrimitivesInput/getEmptyValue");

var _patch = require("../../patch");

var _FormCallbacks = require("../../studio/contexts/FormCallbacks");

var _types = require("@sanity/types");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @alpha
 */
function ArrayOfPrimitivesItem(props) {
  const focusRef = (0, _react.useRef)();
  const {
    member,
    renderItem,
    renderInput
  } = props;
  const {
    onPathBlur,
    onPathFocus,
    onChange
  } = (0, _FormCallbacks.useFormCallbacks)();
  (0, _useDidUpdate.useDidUpdate)(member.item.focused, (hadFocus, hasFocus) => {
    if (!hadFocus && hasFocus) {
      var _focusRef$current;

      (_focusRef$current = focusRef.current) === null || _focusRef$current === void 0 ? void 0 : _focusRef$current.focus();
    }
  });
  const handleBlur = (0, _react.useCallback)(event => {
    onPathBlur(member.item.path);
  }, [member.item.path, onPathBlur]);
  const handleFocus = (0, _react.useCallback)(event => {
    onPathFocus(member.item.path);
  }, [member.item.path, onPathFocus]);
  const handleChange = (0, _react.useCallback)(event => {
    const patches = _patch.PatchEvent.from(event).patches.map(patch => // Map direct unset patches to empty value instead in order to not *remove* elements as the user clears out the value
    // note: this creates the rather "weird" case where the input renders ´0´ when you try to clear it
    patch.path.length === 0 && patch.type === 'unset' ? (0, _patch.set)((0, _getEmptyValue.getEmptyValue)(member.item.schemaType)) : patch);

    onChange(_patch.PatchEvent.from(patches).prefixAll(member.index));
  }, [onChange, member.item.schemaType, member.index]);
  const handleNativeChange = (0, _react.useCallback)(event => {
    let inputValue = event.currentTarget.value;

    if ((0, _types.isNumberSchemaType)(member.item.schemaType)) {
      inputValue = event.currentTarget.valueAsNumber;
    } else if ((0, _types.isBooleanSchemaType)(member.item.schemaType)) {
      inputValue = event.currentTarget.checked;
    } // `valueAsNumber` returns `NaN` on empty input


    const hasEmptyValue = inputValue === '' || typeof inputValue === 'number' && isNaN(inputValue);
    onChange(_patch.PatchEvent.from(hasEmptyValue ? (0, _patch.unset)() : (0, _patch.set)(inputValue)).prefixAll(member.index));
  }, [member.index, member.item.schemaType, onChange]);
  const elementProps = (0, _react.useMemo)(() => ({
    onBlur: handleBlur,
    onFocus: handleFocus,
    id: member.item.id,
    ref: focusRef,
    onChange: handleNativeChange,
    value: String(member.item.value || ''),
    readOnly: Boolean(member.item.readOnly),
    placeholder: member.item.schemaType.placeholder
  }), [handleBlur, handleFocus, handleNativeChange, member.item.id, member.item.readOnly, member.item.schemaType.placeholder, member.item.value]);
  const inputProps = (0, _react.useMemo)(() => {
    return {
      changed: member.item.changed,
      level: member.item.level,
      value: member.item.value,
      readOnly: member.item.readOnly,
      schemaType: member.item.schemaType,
      id: member.item.id,
      path: member.item.path,
      focused: member.item.focused,
      onChange: handleChange,
      validation: member.item.validation,
      presence: member.item.presence,
      elementProps
    };
  }, [member.item.changed, member.item.level, member.item.value, member.item.readOnly, member.item.schemaType, member.item.id, member.item.path, member.item.focused, member.item.validation, member.item.presence, handleChange, elementProps]);
  const renderedInput = (0, _react.useMemo)(() => renderInput(inputProps), [inputProps, renderInput]);
  const onRemove = (0, _react.useCallback)(() => {
    onChange(_patch.PatchEvent.from([(0, _patch.unset)([member.index])]));
  }, [member.index, onChange]);
  const onInsert = (0, _react.useCallback)(event => {
    onChange(_patch.PatchEvent.from([(0, _patch.insert)(event.items, event.position, [member.index])]));
  }, [member.index, onChange]);
  const itemProps = (0, _react.useMemo)(() => {
    return {
      key: member.key,
      index: member.index,
      level: member.item.level,
      value: member.item.value,
      title: member.item.schemaType.title,
      description: member.item.schemaType.description,
      schemaType: member.item.schemaType,
      onInsert,
      onRemove,
      presence: member.item.presence,
      validation: member.item.validation,
      readOnly: member.item.readOnly,
      focused: member.item.focused,
      onFocus: handleFocus,
      inputId: member.item.id,
      path: member.item.path,
      children: renderedInput
    };
  }, [member.key, member.index, member.item.level, member.item.value, member.item.schemaType, member.item.presence, member.item.validation, member.item.readOnly, member.item.focused, member.item.id, member.item.path, onInsert, onRemove, handleFocus, renderedInput]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (0, _react.useMemo)(() => renderItem(itemProps), [itemProps, renderItem]));
}
