// sanity/lib/esm/form/inputs/BooleanInput.js
var _templateObject, _templateObject2;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

import React from 'react';
import styled from 'styled-components';
import { Box, Card, Checkbox, Flex, Switch } from '@sanity/ui';
import { FormFieldHeaderText } from '../components/formField/FormFieldHeaderText';
import { FormFieldStatus } from '../components/formField/FormFieldStatus';
const CenterAlignedBox = styled(Box)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  align-self: center;\n"])));
const ZeroLineHeightBox = styled(Box)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  line-height: 0;\n"])));
export function BooleanInput(props) {
  var _schemaType$options;

  const {
    id,
    value,
    schemaType,
    readOnly,
    elementProps
  } = props;
  const layout = ((_schemaType$options = schemaType.options) === null || _schemaType$options === void 0 ? void 0 : _schemaType$options.layout) || 'switch';
  const indeterminate = typeof value !== 'boolean';
  const checked = value || false;
  const LayoutSpecificInput = layout === 'checkbox' ? Checkbox : Switch;
  return /*#__PURE__*/React.createElement(Card, {
    border: true,
    radius: 1
  }, /*#__PURE__*/React.createElement(Flex, null, /*#__PURE__*/React.createElement(ZeroLineHeightBox, {
    padding: 3
  }, /*#__PURE__*/React.createElement(LayoutSpecificInput, _extends({
    label: schemaType.title
  }, elementProps, {
    indeterminate: indeterminate,
    checked: checked,
    style: {
      margin: -4
    },
    disabled: readOnly
  }))), /*#__PURE__*/React.createElement(Box, {
    flex: 1,
    paddingY: 3
  }, /*#__PURE__*/React.createElement(FormFieldHeaderText, {
    description: schemaType.description,
    inputId: id // validation={validation}
    ,
    title: schemaType.title
  })), /*#__PURE__*/React.createElement(CenterAlignedBox, {
    paddingX: 3,
    paddingY: 1
  }, /*#__PURE__*/React.createElement(FormFieldStatus, {
    maxAvatars: 1,
    position: "top"
  }))));
}
