// sanity/lib/esm/form/inputs/NumberInput.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { TextInput } from '@sanity/ui';
import { getValidationRule } from '../utils/getValidationRule';
export function NumberInput(props) {
  const {
    schemaType,
    validationError,
    value,
    elementProps
  } = props; // Show numpad on mobile if only positive numbers is preferred

  const minRule = getValidationRule(schemaType, 'min');
  const onlyPositiveNumber = ((minRule === null || minRule === void 0 ? void 0 : minRule.constraint) || 0) >= 0;
  return /*#__PURE__*/React.createElement(TextInput, _extends({}, elementProps, {
    type: "number",
    step: "any",
    inputMode: onlyPositiveNumber ? 'numeric' : 'text',
    customValidity: validationError,
    value: value,
    placeholder: schemaType.placeholder,
    pattern: onlyPositiveNumber ? '[d]*' : undefined
  }));
}
