// sanity/lib/esm/form/members/ArrayOfObjectsMembers.js
import React from 'react';
import { ArrayOfObjectsItem } from './items/ArrayOfObjectsItem';
import { MemberItemError } from './MemberItemError';

/**
 * Convenience component for wrapping an array of objects
 * @beta
 */
export function ArrayOfObjectsMembers(props) {
  const {
    members,
    renderInput,
    renderField,
    renderItem,
    renderPreview
  } = props;
  return /*#__PURE__*/React.createElement(React.Fragment, null, members.map(member => {
    if (member.kind === 'item') {
      return /*#__PURE__*/React.createElement(ArrayOfObjectsItem, {
        key: member.key,
        member: member,
        renderInput: renderInput,
        renderField: renderField,
        renderItem: renderItem,
        renderPreview: renderPreview
      });
    }

    if (member.kind === 'error') {
      return /*#__PURE__*/React.createElement(MemberItemError, {
        key: member.key,
        member: member
      });
    } //@ts-expect-error The branching above should cover all possible cases


    console.warn(new Error("Unhandled member kind ".concat(member.kind)));
    return null;
  }));
}
